hazelcast/src/main/java/com/hazelcast/sql/impl/client/SqlClientResult.java

```diff
@@ -18,120 +18,269 @@
     @Override
-    @Nonnull
     public Iterator<SqlRow> iterator() {
-        if (iteratorAccessed) {
+        State state = awaitState();
+
+        ClientIterator iterator = state.iterator;
+
+        if (iterator == null) {
+            throw new IllegalStateException("This result contains only update count");
+        }
+
+        if (iteratorRequested) {
             throw new IllegalStateException("Iterator can be requested only once");
         }
 
-        checkIsRowsResult();
+        iteratorRequested = true;
 
-        iteratorAccessed = true;
         return iterator;
     }
 
     @Override
     public long updateCount() {
-        return updateCount;
+        State state = awaitState();
+
+        return state.updateCount;
     }
 
     @Override
     public void close() {
-        if (iterator == null) {
-            return;
-        }
-
-        try {
-            if (!closed) {
-                if (iterator.last) {
-                    // The last page accessed, so the remote cursor is already closed. No-op.
+        synchronized (mux) {
+            try {
+                // Do nothing if the result is already closed.
+                if (closed) {
                     return;
                 }
 
+                // If the cancellation is initiated before the first response is received, then throw cancellation errors on
+                // the dependent methods (update count, row metadata, iterator).
+                if (state == null) {
+                    onExecuteError(QueryException.cancelledByUser());
+                }
+
+                // Make sure that all subsequent fetches will fail.
+                if (fetch == null) {
+                    fetch = new SqlFetchResult();
+                }
+
+                onFetchFinished(null, QueryException.cancelledByUser());
+
+                // Send the close request.
                 service.close(connection, queryId);
+            } finally {
+                // Set the closed flag to avoid multiple close requests.
+                closed = true;
             }
-        } finally {
+        }
+    }
+
+    /**
+     * Mark result as closed. Invoked when we received an update count or the last page.
+     */
+    private void markClosed() {
+        synchronized (mux) {
             closed = true;
         }
     }
 
-    private void checkIsRowsResult() {
-        if (iterator == null) {
-            throw new IllegalStateException("This result contains only update count");
+    /**
+     * Fetches the next page.
+     */
+    private SqlPage fetch() {
+        synchronized (mux) {
+            // Re-throw previously logged error on successive fetch attempts.
+            if (fetch != null) {
+                assert fetch.getError() != null;
+
+                throw wrap(fetch.getError());
+            }
+
+            // Initiate the fetch.
+            fetch = new SqlFetchResult();
+
+            service.fetchAsync(connection, queryId, cursorBufferSize, this);
```

Comment:
fetchAsync can throw RuntimeException (HazelcastOverloadException is the one that comes to my mind). 
If that happens we are leaving `fetch` field nonnull, which will cause all successive attempts to fail. Looks like exceptions from the invocation system should not necessarily fail successive attempts.

Reply:
SQL doesn't use the invocation subsystem. However, even if it was used, I do not see how a user may react to this. There is a stateful cursor that the user iterates over. In SQL APIs and JDBC, if something goes wrong, a common pattern is to throw an exception and release the resources and then re-execute.

Reply:
I was thinking to enable this kind of usage:
```
 Iterator<SqlRow> iterator = statement.iterator();
        while (iterator.hasNext()) {
            try {
                SqlRow next = iterator.next();
                
            } catch (HazelcastOverloadException e) {
                //Decide on what to do
                //Maybe sleep with exponential backoff! 
            }
        }
```

Similar to the following usage:
```
IMap<Object, Object> map = instance.getMap("map");        
        try {
            Object value = map.getAsync("key");
        } catch (HazelcastOverloadException e) {
        
        }        
```

Note that this exception is thrown only when "hazelcast.client.max.concurrent.invocations" is set. 

It is a special situation that causes an `invoke` to send an exception. Invoke does not throw exceptions otherwise and all the exceptions are delegated to related future.

Reply:
I would not complicate things unless we realize that it is really needed by the user. In my practice, I never saw users doing retries like this why iterating over cursors. In SQL, you typically close the cursor and retry the whole operation. There is even a special exception type for this in JDBC - `SQLTransientException`.

Reply:
To add to the previous comment - SQL doesn't use the invocation subsystem, so it should never see `HazelcastOverloadException` exception anyway.

Reply:
If it is not a usage that you expect, that I am ok with the current implementation.